public/app/plugins/datasource/mixed/MixedDatasource.ts

```diff
@@ -0,0 +1,124 @@
+import cloneDeep from 'lodash/cloneDeep';
+import groupBy from 'lodash/groupBy';
+import map from 'lodash/map';
+import flatten from 'lodash/flatten';
+
+import {
+  DataSourceApi,
+  DataQuery,
+  DataQueryRequest,
+  DataQueryResponse,
+  DataStreamState,
+  DataStreamObserver,
+  DataSourceInstanceSettings,
+} from '@grafana/ui';
+import { LoadingState } from '@grafana/data';
+
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { getProcessedDataFrames } from 'app/features/dashboard/state/PanelQueryState';
+import { toDataQueryError } from 'app/features/dashboard/state/PanelQueryState';
+import { getBackendSrv } from 'app/core/services/backend_srv';
+
+export const MIXED_DATASOURCE_NAME = '-- Mixed --';
+
+export class MixedDatasource extends DataSourceApi<DataQuery> {
+  constructor(instanceSettings: DataSourceInstanceSettings) {
+    super(instanceSettings);
+  }
+
+  async query(request: DataQueryRequest<DataQuery>, observer: DataStreamObserver): Promise<DataQueryResponse> {
+    // Remove any hidden or invalid queries
+    const queries = request.targets.filter(t => {
+      return !t.hide || t.datasource === MIXED_DATASOURCE_NAME;
+    });
+
+    if (!queries.length) {
+      return Promise.resolve({ data: [] }); // nothing
+    }
+
+    // When multiple datasources are used, merge their results
+    // after all queries have finished.  (<=6.3 behavior)
+    const sets = groupBy(queries, 'datasource');
+    const names = Object.keys(sets);
+    if (names.length > 1) {
+      const promises = map(sets, (targets: DataQuery[]) => {
+        const dsName = targets[0].datasource;
+        return getDatasourceSrv()
+          .get(dsName)
+          .then(ds => {
+            const opt = cloneDeep(request);
+            opt.targets = targets;
+            return ds.query(opt);
+          });
+      });
+
+      return Promise.all(promises).then(results => {
+        return { data: flatten(map(results, 'data')) };
+      });
+    }
+
+    // If there is only one datasource used, make multiple requests
+    // and stream the results.  (new in 6.4)
+    const ds = await getDatasourceSrv().get(queries[0].datasource);
+    request.subRequests = [];
```

Comment:
Are we actually modifying the request we got as a param here? To signal something to the caller? If that is the case, I really, really don't like that. Also few questions for the subrequests:
- Is this a one level of subrequests? If so is that some arbitrary limit or just because of how mixed DS works?
- Why subrequests here at all? I would kinda expect the grouping of queries be transparent to the caller. Something along the lines of getting queries, group by DS so they can be sent together so DS can do any optimisation. Ungroup them on the way out so that caller does not need to know about the grouping at all.

Reply:
> What you describe is exactly the behavior you get when there are multiple ...

My main point is more about how this is leaked to the caller. Like caller probably does not need to know this? I think getting a different streams with unique ID is enough and it does not need to know about whether they are hierarchical in some way. Especially that from the UI perspective there isn't really a hierarchy in the queries.

So if this is only internal structuring it is fine (also the ID scheme is fine), but we do `request.subRequests.push(sub);` later on. So this seems again like a mutation of param to signal some info upstream. So apart from problem with param mutation, I think caller does not really need that information anyway. Or is there an usecase when it is necessary?

Reply:
I agree with this, think leaking the concept of subRequests might be unnecessary. And could be an internal detail in the mixed data source. Question is how to communicate the state of individual queries sent in the request to the data source. Maybe we can do this with the refId that is part of the DataFrame result?

Reply:
There is `DataStreamState.key` property so I assume that is the unique identifier of stream for the consumer? `query.refId` is the one that is passed to it here.